domains/client/domain-executor/src/domain_worker.rs

```diff
@@ -106,33 +122,48 @@ pub(crate) async fn handle_block_import_notifications<
                 };
                 // TODO: `.expect()` on `Option` is fine here, but not for `Error`
                 let header = primary_chain_client
-                    .header(
-                        primary_chain_client.hash(block_number)
-                            .expect("Header of imported block must exist; qed")
-                            .expect("Header of imported block must exist; qed")
-                    )
+                    .header(notification.hash)
                     .expect("Header of imported block must exist; qed")
                     .expect("Header of imported block must exist; qed");
+                // TODO: remove fork_choice from BlockInfo
+                let fork_choice = sc_consensus::ForkChoiceStrategy::LongestChain;
                 let block_info = BlockInfo {
                     hash: header.hash(),
                     parent_hash: *header.parent_hash(),
                     number: *header.number(),
                     fork_choice
                 };
-                let _ = block_info_sender.feed(block_info).await;
+                let _ = multi_block_import_sender.feed(BlockImport::Client(block_info)).await;
+            }
+            maybe_subspace_block_import = block_imports.next() => {
+                let (_block_number, _fork_choice, mut block_import_acknowledgement_sender) =
+                    match maybe_subspace_block_import {
+                        Some(block_import) => block_import,
+                        None => {
+                            // Can be None on graceful shutdown.
+                            break;
+                        }
+                    };
+                let _ = multi_block_import_sender.feed(BlockImport::Subspace).await;
```

Comment:
Why not send block info here as well? Block import should be able to short circuit on already imported blocks if the block notification was duplicated? Not sure how much noise this would add though 🤔 in logs and if they would any performace lost.

Reply:
Why sending it if you know it's duplicated for sure? In addition, the notification from sc-consensus-subsubspace is not fired after the finish of the primary block, processing such an notification is unsound in theory, that's why we have this PR to make use of the notification from the natural client instead of sc-consensus-subspace.